test/Main.hs

```diff
@@ -0,0 +1,207 @@
+module Main where
+
+import Control.Monad (forM)
+import qualified Data.ByteString.Lazy as LBS
+import Data.List (isPrefixOf, isSuffixOf, isInfixOf, sort)
+import Data.Ord (Down(..))
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as TE
+import System.Directory
+  ( createDirectoryIfMissing
+  , doesDirectoryExist
+  , doesFileExist
+  , getCurrentDirectory
+  , getModificationTime
+  , getTemporaryDirectory
+  , listDirectory
+  , setCurrentDirectory
+  )
+import System.Exit (ExitCode(..))
+import System.FilePath ((</>), dropExtension, makeRelative, takeDirectory)
+import System.Process (readProcessWithExitCode)
+import Test.Tasty (TestTree, defaultMain, testGroup)
+import Test.Tasty.Golden (goldenVsStringDiff)
+
+-- | Convert a String to a lazy ByteString using UTF-8 encoding.
+stringToLBS :: String -> LBS.ByteString
+stringToLBS = LBS.fromStrict . TE.encodeUtf8 . T.pack
+
+main :: IO ()
+main = do
+  -- Discover the test directory path (relative to the current working directory)
+  cwd <- getCurrentDirectory
+  let testDir = cwd </> "test"
+
+  -- Change to the test directory so that agda2hs can find the agda-lib
+  setCurrentDirectory testDir
+
+  -- Create a temporary build directory
+  tmpDir <- getTemporaryDirectory
+  let buildDir = tmpDir </> "agda2hs-test-build"
+  createDirectoryIfMissing True buildDir
+
+  -- Discover test cases
+  succeedTests <- discoverSucceedTests testDir buildDir
+  failTests <- discoverFailTests testDir buildDir
+
+  -- Run tests
+  defaultMain $ testGroup "agda2hs"
+    [ testGroup "Succeed" succeedTests
+    , testGroup "Fail" failTests
+    ]
+
+-- | Discover all .agda files under the Succeed directory recursively.
+-- Files are ordered by: modification date (newest first), then golden value
+-- modification date (newest first), then alphabetically.
+discoverSucceedTests :: FilePath -> FilePath -> IO [TestTree]
+discoverSucceedTests testDir buildDir = do
+  agdaFiles <- findAgdaFilesRecursive (testDir </> "Succeed")
+  sortedFiles <- sortByModTime agdaFiles (\f -> dropExtension f ++ ".hs")
+  forM sortedFiles $ \agdaFile -> do
+    let testName = dropExtension (makeRelative (testDir </> "Succeed") agdaFile)
+        goldenFile = dropExtension agdaFile ++ ".hs"
+    return $ succeedTest testDir buildDir testName agdaFile goldenFile
+
+-- | Discover all .agda files under the Fail directory.
+-- Files are ordered by: modification date (newest first), then golden value
+-- modification date (newest first), then alphabetically.
+discoverFailTests :: FilePath -> FilePath -> IO [TestTree]
```

Comment:
There is a lot of code duplication between `discoverSucceedTests` and `discoverFailTests`. Could you factor out a more general function `discoverTests` that takes the name of the directory ("Succeed" or "Fail"), the extension of the golden files (".hs" or ".err") and the testing function (`succeedTest` or `failTest`) as inputs, and make these two functions call the general one?